File: src/main/java/KeyCombo.java
import processing.core.PApplet;
import processing.event.KeyEvent;

import java.util.Objects;


enum ValidOS {
    ANY,
    MAC,
    NON_MAC
}

public class KeyCombo {
    static ValidOS currentOS = determineCurrentOS();
    private final int keyCode;
    private final int modifiers;
    private final ValidOS validOS;
    public KeyCombo(int keyCode) {
        this(keyCode, 0, ValidOS.ANY);
    }
    public KeyCombo(int keyCode, int modifiers, ValidOS validOS) {
        this.keyCode = Character.isLowerCase(keyCode) ? Character.toUpperCase(keyCode) : keyCode;
        this.modifiers = modifiers;
        this.validOS = validOS;
    }

    public KeyCombo(int keyCode, int modifiers) {
        this(keyCode, modifiers, ValidOS.ANY);
    }

    public KeyCombo(char keyCode, int modifiers) {
        this(keyCode, modifiers, ValidOS.ANY);
    }

    private static ValidOS determineCurrentOS() {
        String osName = System.getProperty("os.name").toLowerCase();
        if (osName.contains("mac")) {
            return ValidOS.MAC;
        } else {
            return ValidOS.NON_MAC;
        }
    }

    // use this for testing purposes only
    static void setCurrentOS(ValidOS os) {
        currentOS = os;
    }

    public ValidOS getValidOS() {
        return validOS;
    }

    // used at runtime to make sure that the incoming keyEvent matches a particular KeyCombo
    public boolean matches(KeyEvent event) {
        return this.getKeyCode() == event.getKeyCode() &&
                this.getModifiers() == event.getModifiers() &&
                isValidForCurrentOS();
    }

    public int getKeyCode() {
        return keyCode;
    }

    public int getModifiers() {
        return modifiers;
    }

    public boolean isValidForCurrentOS() {

        return switch (validOS) {
            case ANY -> true;
            case MAC -> currentOS == ValidOS.MAC;
            case NON_MAC -> currentOS == ValidOS.NON_MAC;
        };
    }

    @Override
    public int hashCode() {
        return Objects.hash(keyCode, modifiers, validOS);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        KeyCombo keyCombo = (KeyCombo) o;
        return keyCode == keyCombo.keyCode &&
                modifiers == keyCombo.modifiers &&
                validOS == keyCombo.validOS;
    }

    @Override
    public String toString() {
        // Special case for Shift+=, display as +
        if ((modifiers & KeyEvent.SHIFT) != 0 && keyCode == '=') {
            return "+";
        }

        StringBuilder keyTextBuilder = new StringBuilder();

        if ((modifiers & KeyEvent.META) != 0) {
            keyTextBuilder.append("⌘");
        }
        if ((modifiers & KeyEvent.CTRL) != 0) {
            keyTextBuilder.append("^");
        }
        if ((modifiers & KeyEvent.SHIFT) != 0) {
            keyTextBuilder.append("↑");
        }
        if ((modifiers & KeyEvent.ALT) != 0) {
            keyTextBuilder.append("⌥");
        }

        switch (keyCode) {
            case PApplet.UP -> keyTextBuilder.append("↑");
            case PApplet.DOWN -> keyTextBuilder.append("↓");
            case PApplet.LEFT -> keyTextBuilder.append("←");
            case PApplet.RIGHT -> keyTextBuilder.append("→");
            case 32 -> keyTextBuilder.append("Space");
            default -> keyTextBuilder.append((char) keyCode);
        }

        return keyTextBuilder.toString();
    }
}

File: src/main/kotlin/actions/KeyCombo.kt
package actions

import processing.core.PApplet
import processing.event.KeyEvent

class KeyCombo @JvmOverloads constructor(private val keyCode: Int, private val modifiers: Int = 0, private val validOS: ValidOS = ValidOS.ANY) {

    // we only want to do the conversion once so cache it in the val
    // we do it this way so that we can make this a data class and still have val semantics
    // then the dataclass doesn't need to create an explicit hashCode
    private val cachedKeyCode: Int = if (Character.isLowerCase(keyCode)) Character.toUpperCase(keyCode) else keyCode
    private val processingKeyCode: Int
        get() = cachedKeyCode

    constructor(keyCode: Char, modifiers: Int) : this(keyCode.code, modifiers, ValidOS.ANY)

    // used at runtime to make sure that the incoming keyEvent matches a particular actions.KeyCombo
    fun matches(event: KeyEvent): Boolean {
        return processingKeyCode == event.keyCode && modifiers == event.modifiers &&
                isValidForCurrentOS
    }

    val isValidForCurrentOS: Boolean
        get() = when (validOS) {
            ValidOS.ANY -> true
            ValidOS.MAC -> currentOS === ValidOS.MAC
            ValidOS.NON_MAC -> currentOS === ValidOS.NON_MAC
        }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val keyCombo = other as KeyCombo
        return processingKeyCode == keyCombo.processingKeyCode && modifiers == keyCombo.modifiers && validOS === keyCombo.validOS
    }

    // later:  “␛” or better yet: ⎋ - but large enough to see - use it for closing info panels
    override fun toString(): String {

        val keyTextBuilder = StringBuilder()

        when {
            modifiers and KeyEvent.META != 0 -> keyTextBuilder.append("⌘")
            modifiers and KeyEvent.CTRL != 0 -> keyTextBuilder.append("^")
            modifiers and KeyEvent.SHIFT != 0 -> keyTextBuilder.append("⇧")
            modifiers and KeyEvent.ALT != 0 -> keyTextBuilder.append("⌥")
        }

        when {
            //Special case for Shift+=, display as +
            modifiers and KeyEvent.SHIFT != 0 && processingKeyCode == '='.code -> keyTextBuilder.append("+")
            processingKeyCode == PApplet.UP -> keyTextBuilder.append("↑")
            processingKeyCode == PApplet.DOWN -> keyTextBuilder.append("↓")
            processingKeyCode == PApplet.LEFT -> keyTextBuilder.append("←")
            processingKeyCode == PApplet.RIGHT -> keyTextBuilder.append("→")
            processingKeyCode == 32 -> keyTextBuilder.append("Space")
            else -> keyTextBuilder.append(processingKeyCode.toChar())
        }

        return keyTextBuilder.toString()
    }

    override fun hashCode(): Int {
        var result = keyCode
        result = 31 * result + modifiers
        result = 31 * result + validOS.hashCode()
        return result
    }

    companion object {
        val currentOS = determineCurrentOS()

        private fun determineCurrentOS(): ValidOS {
            val osName = System.getProperty("os.name").lowercase()
            return if (osName.contains("mac")) {
                ValidOS.MAC
            } else {
                ValidOS.NON_MAC
            }
        }
    }
}